### PR TITLE
[core][autoscaler] ray status - integration with feature flag controlled by server [4/x] 

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
This PR: 
- Implements feature flag that's stored in GCS kv so that client (ray status, specifically) would alter behaviour based on the flag passed in when initializing the head with `RAY_enable_autoscaler_v2`
- Tests parity with existing ray status 

Also tested with dashboard (starting a cluster with feature flag on and dashboard is working). 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rickyyx** [comment]
Depends on https://github.com/ray-project/ray/pull/37406

File: BUILD.bazel
Comment (rickyyx) [inline BUILD.bazel:2594]: This is actually much easier than what I did initially (adding another build target which put things under `ray/core/generated/experimental`)

File: python/setup.py
Comment (rickyyx) [inline python/setup.py:187]: Need code owner stamp.

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_server.cc:668]: Should we crash here?

**Review by fishbone — COMMENTED** [review]
LG.

**Review by fishbone — APPROVED** [review]

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (rickyyx) [reply src/ray/gcs/gcs_server/gcs_server.cc:668]: maybe we could eventually when we roll out the feature. I am a bit concerned this codepath gets triggered and cause behaviour change when this put failed before we flip the feature flag... 

So if it fails, it will behave as if autoscaler v2 not enabled to client.

File: python/setup.py
Comment (rickyyx) [inline python/setup.py:187]: This reverts https://github.com/ray-project/ray/pull/37403 since it's no longer needed.

**Review by richardliaw — APPROVED** [review]

File: dashboard/modules/reporter/reporter_head.py
Comment (rkooo567) [inline dashboard/modules/reporter/reporter_head.py:81]: Does this already pre-process pg resources btw? (that's prerequisite before returning the structured data via this API)

File: python/ray/autoscaler/v2/utils.py
Comment (rkooo567) [inline python/ray/autoscaler/v2/utils.py:527]: Isn't this too expensive? why don't we just follow the previous pattern?

Comment (rkooo567) [reply python/ray/autoscaler/v2/utils.py:527]: More concretely, I am concerned about the future (that people overuse this API)

Comment (rkooo567) [reply python/ray/autoscaler/v2/utils.py:527]: maybe we can keep the state in memory after the first call?

File: dashboard/modules/reporter/reporter_head.py
Comment (rickyyx) [reply dashboard/modules/reporter/reporter_head.py:81]: wdym by `pre-process` pg resources?

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (rickyyx) [reply src/ray/gcs/gcs_server/gcs_server.cc:668]: if this fails, it's default v1. From server's perspective, v1/v2 actually doesn't make a difference, it's just v2 will support a few more rpc calls. 

But sure, I am open to crashing here.

File: python/ray/autoscaler/v2/utils.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/utils.py:533]: This method would be removed as soon as we migrate to autoscaler v2.

The callers right now are not in any critical path, mainly for observability and sdks.

Previous pattern is not ideal since client would also need to specify the env variable. (and it might be inconsistent with the server)

Comment (rickyyx) [reply python/ray/autoscaler/v2/utils.py:533]: The caching is a bit tricky, i tried caching in the module (having a global variable outside of the function), but this would interfere with multiple ray sessions (e.g. ray.init(v2) -> ray.shutdown() -> ray.init(v1), would have cache value. 

We could not cache it on the global ray driver also because one current usecase (in the reporter head) does not have driver initialized.

**Comment by rickyyx** [comment]
Updates: 
- Added many integration tests. 
- Crash GCS if feature flag not stored properly.

File: dashboard/modules/reporter/reporter_head.py
Comment (rkooo567) [reply dashboard/modules/reporter/reporter_head.py:81]: ah I meant the pg resources need to be parsed and structured (e.g., CPU_group_id_... things)

File: python/ray/autoscaler/v2/utils.py
Comment (rkooo567) [reply python/ray/autoscaler/v2/utils.py:533]: can you comment me which APIs use this flag under the hood? Let's just make sure we don't have excessive calls here.

**Review by rkooo567 — APPROVED** [review]
LGTM. Can we just follow up with https://github.com/ray-project/ray/pull/37473#discussion_r1268812000 before merge? Or we can merge and discuss.

File: python/ray/autoscaler/v2/utils.py
Comment (rickyyx) [reply python/ray/autoscaler/v2/utils.py:533]: Right now, 

1. debug_status (from `ray status`): https://github.com/ray-project/ray/blob/1a1fdeb68329839c225b7136c983689bae70525e/python/ray/autoscaler/_private/commands.py#L115-L127
2. debug_status (from reporter head's handling of `/api/cluster_events`): https://github.com/ray-project/ray/blob/3d4e4ab9636f7ad181b0f73c829699a5c3c1fe0b/dashboard/modules/reporter/reporter_head.py#L116
3. request_resources: https://github.com/ray-project/ray/blob/1a1fdeb68329839c225b7136c983689bae70525e/python/ray/autoscaler/_private/commands.py#L183-L185

File: dashboard/modules/reporter/reporter_head.py
Comment (rickyyx) [reply dashboard/modules/reporter/reporter_head.py:81]: yeah, it's parsed (by the legacy code). So nothing is changed here as of now.

File: python/ray/autoscaler/v2/utils.py
Comment (rkooo567) [reply python/ray/autoscaler/v2/utils.py:533]: hmm the dashboard part is concerning because the internal kv get can block the thread for 30+ seconds in a large cluster. 

is there a way to make it async for that part?

Comment (rkooo567) [reply python/ray/autoscaler/v2/utils.py:533]: > but this would interfere with multiple ray sessions (e.g. ray.init(v2) -> ray.shutdown() -> ray.init(v1), would have cache value.

Also feel like this is not a problem? There's no use case you use autoscaler V1 in first ray init and V2 in the second ray.init within the same python script?

Comment (rickyyx) [reply python/ray/autoscaler/v2/utils.py:533]: > hmm the dashboard part is concerning because the internal kv get can block the thread for 30+ seconds in a large cluster.

The rest of the code (e.g. getting debug string) is already using internal kv with blocking. So it would already be bad. 

> Also feel like this is not a problem? There's no use case you use autoscaler V1 in first ray init and V2 in the second ray.init within the same python script?

Well, this happens in test. But I guess I will just rewrite the test to circumvent this.

**Pull request merged**
